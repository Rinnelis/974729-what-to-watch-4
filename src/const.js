export const SIMILAR_FILMS_AMOUNT = 4;

export const MAX_SHOWN_FILMS_AMOUNT = 8;

export const MAX_GENRES_AMOUNT = 10;

export const ALL_GENRES = `All genres`;

export const Time = {
  SECONDS_PER_MINUTE: 60,
  MINUTES_PER_HOUR: 60
};

export const RatingLevel = {
  AWESOME: `Awesome`,
  VERY_GOOD: `Very good`,
  GOOD: `Good`,
  NORMAL: `Normal`,
  BAD: `Bad`
};

export const MovieNav = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const AuthStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`
};

export const Page = {
  MAIN: `/`,
  SIGN_IN: `/login`,
  FILM: `/films`,
  ADD_REVIEW: `/review`,
  MY_LIST: `/mylist`,
  PLAYER: `/player`,
};

export const ValidReview = {
  MIN: 50,
  MAX: 400,
};

export const MIN_RATING = `1`;

export const Error = {
  UNAUTHORIZED: 401
};

export const ENTRY_POINT = `https://4.react.pages.academy/wtw`;

export const REQUEST_TIMEOUT = 5000;
