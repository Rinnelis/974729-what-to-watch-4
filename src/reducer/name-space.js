export const NameSpace = {
  DATA: `DATA`,
  MOVIES: `MOVIES`,
  USER: `USER`,
};
