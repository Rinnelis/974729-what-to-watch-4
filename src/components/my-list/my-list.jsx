import React from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {ProjectPropTypes} from "../../project-prop-types.js";
import MoviesList from "../movies-list/movies-list.jsx";
import Footer from "../footer/footer.jsx";
import {Page} from "../../const.js";
import {getFavoriteFilms, getFavoriteFilmsStatus} from "../../reducer/data/selectors.js";
import {Operation} from "../../reducer/data/data.js";
import {getUserData} from "../../reducer/user/selectors.js";
import {ActionCreator} from "../../reducer/movies/movies.js";

const MyList = (props) => {
  const {user, favoriteFilms, onMovieChoose, isLoadingFavoriteFilms} = props;
  const {name, avatarUrl} = user;

  const getFavoriteFilmsMsg = () => {
    if (isLoadingFavoriteFilms.isLoadingFavoriteFilms && !isLoadingFavoriteFilms.loadFavoriteFilmsError) {
      return `Favorite films are loading...`;
    } else if (isLoadingFavoriteFilms.isLoadingFavoriteFilms && isLoadingFavoriteFilms.loadFavoriteFilmsError) {
      return `Server error occurred, please try again later`;
    }

    return false;
  };

  return (
    <React.Fragment>
      <div className="user-page">
        <header className="page-header user-page__head">
          <div className="logo">
            <Link to={Page.MAIN} className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <h1 className="page-title user-page__title">My list</h1>

          <div className="user-block">
            <div className="user-block__avatar">
              <Link to={Page.MAIN}>
                <img src={avatarUrl} alt={name} width="63" height="63" />
              </Link>
            </div>
          </div>
        </header>

        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          {getFavoriteFilmsMsg() ||
          <MoviesList
            films={favoriteFilms}
            onCardClick={onMovieChoose}
          />
          }
        </section>

        <Footer />
      </div>
    </React.Fragment>
  );
};

MyList.propTypes = {
  user: ProjectPropTypes.USER,
  favoriteFilms: PropTypes.arrayOf(ProjectPropTypes.FILM).isRequired,
  onMovieChoose: PropTypes.func.isRequired,
  isLoadingFavoriteFilms: PropTypes.shape({
    isLoadingFavoriteFilms: PropTypes.bool.isRequired,
    loadFavoriteFilmsError: PropTypes.bool.isRequired,
  }),
};

const mapStateToProps = (state) => ({
  user: getUserData(state),
  favoriteFilms: getFavoriteFilms(state),
  isLoadingFavoriteFilms: getFavoriteFilmsStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onMovieChoose(film) {
    dispatch(ActionCreator.chooseMovie(film));
    dispatch(Operation.loadComments(film.id));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(MyList);
