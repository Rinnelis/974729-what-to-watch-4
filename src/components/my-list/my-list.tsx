import * as React from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {Films, User} from "../../types";
import MoviesList from "../movies-list/movies-list";
import Footer from "../footer/footer";
import {Page} from "../../const";
import {getFavoriteFilms, getFavoriteFilmsStatus} from "../../reducer/data/selectors";
import {Operation} from "../../reducer/data/data";
import {getUserData} from "../../reducer/user/selectors";

interface Props {
  user: User;
  favoriteFilms: Films;
  isLoadingFavoriteFilms: {
    isLoadingFavoriteFilms: boolean;
    loadFavoriteFilmsError: boolean;
  };
  loadFavoriteFilms: () => void;
}

class MyList extends React.PureComponent<Props> {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {loadFavoriteFilms} = this.props;
    loadFavoriteFilms();
  }

  render() {
    const {user, favoriteFilms, isLoadingFavoriteFilms} = this.props;
    const {name, avatarUrl} = user;

    const getFavoriteFilmsMsg = () => {
      if (isLoadingFavoriteFilms.isLoadingFavoriteFilms && !isLoadingFavoriteFilms.loadFavoriteFilmsError) {
        return `Favorite films are loading...`;
      } else if (isLoadingFavoriteFilms.isLoadingFavoriteFilms && isLoadingFavoriteFilms.loadFavoriteFilmsError) {
        return `Server error occurred, please try again later`;
      }

      return false;
    };

    return (
      <React.Fragment>
        <div className="user-page">
          <header className="page-header user-page__head">
            <div className="logo">
              <Link to={Page.MAIN} className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </Link>
            </div>

            <h1 className="page-title user-page__title">My list</h1>

            <div className="user-block">
              <div className="user-block__avatar">
                <Link to={Page.MAIN}>
                  <img src={avatarUrl} alt={name} width="63" height="63" />
                </Link>
              </div>
            </div>
          </header>

          <section className="catalog">
            <h2 className="catalog__title visually-hidden">Catalog</h2>

            {getFavoriteFilmsMsg() ||
            <MoviesList
              films={favoriteFilms}
            />
            }
          </section>

          <Footer />
        </div>
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => ({
  user: getUserData(state),
  favoriteFilms: getFavoriteFilms(state),
  isLoadingFavoriteFilms: getFavoriteFilmsStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadFavoriteFilms() {
    dispatch(Operation.loadFavoriteFilms());
  },
});

export {MyList};
export default connect(mapStateToProps, mapDispatchToProps)(MyList);
