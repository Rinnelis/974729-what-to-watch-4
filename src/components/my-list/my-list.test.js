import React from "react";
import {Router} from "react-router-dom";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import {MyList} from "./my-list.jsx";
import {user, films, film} from "../../test-data.js";
import {NameSpace} from "../../reducer/name-space.js";
import {AuthStatus} from "../../const.js";
import history from "../../history.js";

const mockStore = configureStore([]);

const store = mockStore({
  [NameSpace.USER]: {
    authStatus: AuthStatus.AUTH,
    authError: false,
    user,
    isAuthInProgress: false,
  },
  [NameSpace.DATA]: {
    film,
    films,
    favoriteFilms: films,
    isLoadingFavoriteFilms: false,
    loadFavoriteFilmsError: false,
  }
});

it(`Should render MyList`, () => {
  const tree = renderer.create(
      <Router history={history}>
        <Provider store={store}>
          <MyList
            user={user}
            favoriteFilms={films}
            loadFavoriteFilms={() => {}}
            isLoadingFavoriteFilms={{
              isLoadingFavoriteFilms: false,
              loadFavoriteFilmsError: false,
            }}
          />
        </Provider>
      </Router>, {
        createNodeMock: () => {
          return {};
        }
      }
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

