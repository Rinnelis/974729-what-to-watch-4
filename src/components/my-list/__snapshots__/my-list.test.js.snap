// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render MyList 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          <img
            alt="Oliver.conner"
            height="63"
            src="img/1.png"
            width="63"
          />
        </a>
      </div>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/2"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Fantastic Beasts: The Crimes of Grindelwald"
              height="175"
              src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/2"
            onClick={[Function]}
          >
            Fantastic Beasts: The Crimes of Grindelwald
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/3"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Shutter Island"
              height="175"
              src="img/shutter-island.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/3"
            onClick={[Function]}
          >
            Shutter Island
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/4"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Macbeth"
              height="175"
              src="img/macbeth.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/4"
            onClick={[Function]}
          >
            Macbeth
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/5"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Aviator"
              height="175"
              src="img/aviator.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/5"
            onClick={[Function]}
          >
            Aviator
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/6"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="We need to talk about Kevin"
              height="175"
              src="img/we-need-to-talk-about-kevin.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/6"
            onClick={[Function]}
          >
            We need to talk about Kevin
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/7"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="What We Do in the Shadows"
              height="175"
              src="img/what-we-do-in-the-shadows.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/7"
            onClick={[Function]}
          >
            What We Do in the Shadows
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/8"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Revenant"
              height="175"
              src="img/revenant.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/8"
            onClick={[Function]}
          >
            Revenant
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/9"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Johnny English"
              height="175"
              src="img/johnny-english.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/9"
            onClick={[Function]}
          >
            Johnny English
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/10"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Pulp Fiction"
              height="175"
              src="img/pulp-fiction.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/10"
            onClick={[Function]}
          >
            Pulp Fiction
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/11"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="No Country for Old Men"
              height="175"
              src="img/no-country-for-old-men.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/11"
            onClick={[Function]}
          >
            No Country for Old Men
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/12"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Snatch"
              height="175"
              src="img/snatch.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/12"
            onClick={[Function]}
          >
            Snatch
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          href="/films/13"
          onClick={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <video
              className="player__video"
            />
            <img
              alt="Bohemian Rhapsody"
              height="175"
              src="img/bohemian-rhapsody.jpg"
              width="280"
            />
          </div>
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/13"
            onClick={[Function]}
          >
            Bohemian Rhapsody
          </a>
        </h3>
      </article>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
